### PR TITLE
L2-224: NNSDappCanister class

### PR DESCRIPTION
**Pull request by lmuntanerDfinity**
# Motivation

Create NNSDappCanister service to interact with the NNS Canister.

# Changes

* Add NNSDappCanister service

# Tests

* nnsDappCanister.spec.ts

### CODE REVIEW

File: frontend/svelte/jest-setup.ts [hunk 1]
```diff
@@ -4,6 +4,9 @@ import "@testing-library/jest-dom";
 import { TextDecoder, TextEncoder } from "util";
 import { IntersectionObserverPassive } from "./src/tests/mocks/infinitescroll.mock";
 
+// Global Mocks
+// Needed for HttpAgent
+global.fetch = jest.fn();
```
Comment (peterpeterparker) [inline frontend/svelte/jest-setup.ts:9]: are you in sync with main branch ? `createAgent` is now mocked for any tests therefore it should not be needed

File: frontend/svelte/src/lib/canisters/nnsDapp/nnsDappCanister.ts [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+export class NNSDappCanister {
+  private constructor(
+    private readonly service: NNSDappService,
+    private readonly certifiedService: NNSDappService
+  ) {
+    this.service = service;
+    this.certifiedService = certifiedService;
+  }
+
+  public static create(options: NNSDappCanisterOptions) {
```
Comment (peterpeterparker) [inline frontend/svelte/src/lib/canisters/nnsDapp/nnsDappCanister.ts:20]: - to follow our common pattern, I would make the `options: GovernanceCanisterOptions = {` optional

- if optional, then properties can be destructed

File: frontend/svelte/src/lib/canisters/nnsDapp/nnsDappCanister.ts [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+        canisterId,
+      });
+
+    return new NNSDappCanister(service, certifiedService);
+  }
+
+  /**
+   * Used to be able to create subAccounts
+   *
+   * TODO: Understand why we need this?
```
Comment (peterpeterparker) [inline frontend/svelte/src/lib/canisters/nnsDapp/nnsDappCanister.ts:44]: there is a JIRA to track the todo?

File: frontend/svelte/src/lib/canisters/nnsDapp/nns_dapp.update.sh [hunk 4]
```diff
@@ -0,0 +1,13 @@
+#!/usr/bin/env bash
```
Comment (peterpeterparker) [inline frontend/svelte/src/lib/canisters/nnsDapp/nns_dapp.update.sh:1]: not sure that's the best place for that script. maybe better under `scripts` or at root of the project?

File: frontend/svelte/src/lib/canisters/nnsDapp/nns_dappTypes.d.ts [hunk 5]
```diff
@@ -0,0 +1,187 @@
+import type { Principal } from "@dfinity/principal";
```
Comment (peterpeterparker) [inline frontend/svelte/src/lib/canisters/nnsDapp/nns_dappTypes.d.ts:1]: - if you use `import` and `export` it should not be a `.d.ts` but rather a `.ts` file

File: frontend/svelte/src/tests/lib/canisters/nnsDappCanister.spec.ts [hunk 6]
```diff
@@ -0,0 +1,97 @@
...
+import { mock } from "jest-mock-extended";
+import { NNSDappCanister } from "../../../lib/canisters/nnsDapp/nnsDappCanister";
+import type { NNSDappService } from "../../../lib/canisters/nnsDapp/nns_dapp.idl";
+import type {
+  CreateSubAccountResponse,
+  SubAccountDetails,
+} from "../../../lib/canisters/nnsDapp/nns_dappTypes";
+
+describe("NNSDapp.createSubAccount", () => {
+  const defaultAgent = new HttpAgent({ host: "http://localhost:8000" });
```
Comment (peterpeterparker) [inline frontend/svelte/src/tests/lib/canisters/nnsDappCanister.spec.ts:12]: once the properties optional you might not need this

**Review by peterpeterparker — CHANGES_REQUESTED** [review]
Regarding naming, it seems we do not have uppercase in file names. Can you go through the folders and file names of this PR an align the names?

For example: `frontend/svelte/src/lib/canisters/nnsDapp/nns_dappTypes.d.ts` -> 

- `nnsDapp` maybe `nns-dapp` as project name
- `nns_dappTypes`  as `nns-dapp.types.ts`

etc.

File: frontend/svelte/src/lib/canisters/nnsDapp/nnsDappCanister.ts (hunk 3 above)
Comment (lmuntanerDfinity) [reply frontend/svelte/src/lib/canisters/nnsDapp/nnsDappCanister.ts:44]: not yet

File: frontend/svelte/src/lib/canisters/nnsDapp/nnsDappCanister.ts (hunk 2 above)
Comment (lmuntanerDfinity) [reply frontend/svelte/src/lib/canisters/nnsDapp/nnsDappCanister.ts:20]: ok

File: frontend/svelte/src/lib/canisters/nnsDapp/nns_dapp.update.sh (hunk 4 above)
Comment (lmuntanerDfinity) [reply frontend/svelte/src/lib/canisters/nnsDapp/nns_dapp.update.sh:1]: it's used very specific for this. Maybe script inside the `svelte` folder?

File: frontend/svelte/src/lib/canisters/nnsDapp/nns_dappTypes.d.ts (hunk 5 above)
Comment (lmuntanerDfinity) [reply frontend/svelte/src/lib/canisters/nnsDapp/nns_dappTypes.d.ts:1]: ok

File: frontend/svelte/src/tests/lib/canisters/nnsDappCanister.spec.ts (hunk 6 above)
Comment (lmuntanerDfinity) [reply frontend/svelte/src/tests/lib/canisters/nnsDappCanister.spec.ts:12]: ok

File: frontend/svelte/src/lib/canisters/nnsDapp/nnsDappCanister.ts (hunk 2 above)
Comment (lmuntanerDfinity) [reply frontend/svelte/src/lib/canisters/nnsDapp/nnsDappCanister.ts:20]: But it's not optional, we need the agent and canister id. Shall I add the agent and canister id as first params and then optional the `serviceOverride` and `certifiedServiceOverride`?

File: frontend/svelte/src/lib/canisters/nnsDapp/nnsDappCanister.ts (hunk 3 above)
Comment (lmuntanerDfinity) [reply frontend/svelte/src/lib/canisters/nnsDapp/nnsDappCanister.ts:44]: do we know who we can ask?

File: frontend/svelte/src/lib/canisters/nnsDapp/nnsDappCanister.ts (hunk 2 above)
Comment (lmuntanerDfinity) [reply frontend/svelte/src/lib/canisters/nnsDapp/nnsDappCanister.ts]: https://dfinity.atlassian.net/browse/L2-299

File: frontend/svelte/src/lib/canisters/nnsDapp/nnsDappCanister.ts (hunk 3 above)
Comment (lmuntanerDfinity) [reply frontend/svelte/src/lib/canisters/nnsDapp/nnsDappCanister.ts]: ask during NNS Intro

**Review by peterpeterparker — APPROVED** [review]

**Pull request merged**